backend/pubsub.js

```diff
@@ -1,127 +1,295 @@
-              spark.leave(contractID, () => done(success))
-              break
-            case PUB:
-              spark.room(contractID).except(id).write(req)
-              break
-            default:
-              console.error(bold.red(`[pubsub] client ${id} didn't give us a valid type!`), req)
-              spark.leaveAll()
-              done(reply(ERROR, `invalid type: ${type}`))
-              spark.end()
+          if (defaultHandler) {
+            defaultHandler.call(socket, ...args)
           }
-        } catch (err) {
-          logger(err)
-          done(reply(ERROR, err))
+          if (customHandler) {
+            customHandler.call(socket, ...args)
+          }
+        } catch (error) {
+          logger(error)
         }
       })
-      spark.on('leaveallrooms', (rooms) => {
-        console.log(bold.yellow(`[pubsub] ${id} leaveallrooms`))
-        // this gets called on spark.leaveAll and 'disconnection'
-        rooms.forEach(contractID => {
-          primus.room(contractID).write(reply(UNSUB, { contractID, id }))
-        })
-      })
-
-      spark.on('data', function (data) {
-        console.log(bold.red('[pubsub] received UNHANDLED DATA from client:', data))
-      })
     })
+  },
+  error (error: Error) {
+    console.log('[pubsub] Server event: error', error)
+  },
+  headers () {
+  },
+  listening () {
+    console.log('[pubsub] Server event: listening')
+  }
+}
 
-    primus.on('disconnection', function (spark) {
-      // the spark that disconnected
-      console.log(bold.yellow(`[pubsub] ${spark.id} disconnection`))
-    })
-  })
+// Default handlers for server-side client socket events.
+// The `this` binding refers to the connected `ws` socket object.
+const defaultSocketEventHandlers = {
+  close (code: string, reason: string) {
+    const socketID = this.id
+    const { server } = this
+    // Notify other client sockets that this one has left any room they shared.
+    for (const contractID of this.subscriptions) {
+      const subscribers = server.subscribersByContractID[contractID]
+      // Remove this socket from the subscribers of the given contract.
+      subscribers.delete(this)
+      const notification = createNotification(UNSUB, { contractID, socketID })
+      server.broadcast(notification, { to: subscribers })
+    }
+    this.subscriptions.clear()
+  },
+
+  message (data: string) {
+    let msg: Message = { type: '' }
+    try {
+      msg = messageParser(data)
+    } catch (error) {
+      console.error(bold.red(`[pubsub] Malformed message: ${error.message}`))
+      return this.terminate()
+    }
+    this.activeSinceLastPing = true
+    const handler = this.server.messageHandlers[msg.type]
+
+    if (handler) {
+      try {
+        handler.call(this, msg)
+      } catch (error) {
+        // Log the error message and stack trace but do not send it to the client.
+        logger(error)
+        // Should we call 'this.terminate()' instead?
```

Comment:
What are the tradeoffs?

Reply:
According to the author in https://github.com/websockets/ws/issues/1596:
> ws.terminate() immediately destroys the connection, ws.close() waits for close timer to expire before destroying the connection if the other peer is not responsive.

So in my understanding, `terminate()` call `close()` internally, but doesn't wait before removing event listeners and releasing memory. Some investigation is needed to see how this affects automatic reconnection of the client.

Reply:
> So in my understanding, terminate() call close()

Based on the text you quoted, wouldn't it be the other way around (`close()` calling `terminate()` internally)?

Reply:
`close()` does not need to remove all event listeners and clear references like terminate() does, so it should not call it I guess.

Reply:
Ah, I see